rpc/types/src/json.rs

```diff
@@ -0,0 +1,127 @@
+use crate::{
+    base::{EmptyRequestBase, EmptyResponseBase, ResponseBase},
+    macros::define_request_and_response,
+};
+
+//---------------------------------------------------------------------------------------------------- Struct definitions
+// This generates 2 structs:
+//
+// - `GetBlockTemplateRequest`
+// - `GetBlockTemplateResponse`
+//
+// with some interconnected documentation.
+define_request_and_response! {
+    // The markdown tag for Monero RPC documentation. Not necessarily the endpoint.
+    get_block_template,
+
+    // The commit hash and `$file.$extension` in which this type is defined in
+    // the Monero codebase in the `rpc/` directory, followed by the specific lines.
+    cc73fe71162d564ffda8e549b79a350bca53c454 => core_rpc_server_commands_defs.h => 943..=994,
+
+    // The base type name.
+    GetBlockTemplate,
+
+    // The base request type.
+    //
+    // This must be a type found in [`crate::base`].
+    // It acts as a "base" that gets flattened into
+    // the actually request type.
+    //
+    // "Flatten" means the field(s) of a struct gets inlined
+    // directly into the struct during (de)serialization, see:
+    // <https://serde.rs/field-attrs.html#flatten>.
+    //
+    // For example here, we're using [`crate::base::EmptyRequestBase`],
+    // which means that there is no extra fields flattened.
+    //
+    // If a request is not specified here, it will create a `type alias YOUR_REQUEST_TYPE = ()`
+    // instead of a `struct`, see below in other macro definitions for an example.
+    EmptyRequestBase {
+        reserve_size: u64,
+        wallet_address: String,
+    },
+
+    // The base response type.
+    //
+    // This is the same as the request base type,
+    // it must be a type found in [`crate::base`].
+    //
+    // If there are any additional attributes (`/// docs` or `#[derive]`s)
+    // for the struct, they go here, e.g.:
+    // #[derive(Copy)]
+    ResponseBase {
+        // This is using `crate::base::ResponseBase`,
+        // so the type we generate will contain this field:
+        // ```
+        // base: crate::base::ResponseBase,
+        // ```
+        //
+        // This is flattened with serde and epee, so during
+        // (de)serialization, it will act as if there are 2 extra fields here:
+        // ```
+        // status: crate::Status,
+        // untrusted: bool,
+        // ```
+
+        // Within the `{}` is an infinite matching pattern of:
+        // ```
+        // $ATTRIBUTES
+        // $FIELD_NAME: $FIELD_TYPE,
+        // ```
+        // The struct generated and all fields are `pub`.
+        difficulty: u64,
+        wide_difficulty: String,
+        difficulty_top64: u64,
+        height: u64,
+        reserved_offset: u64,
+        expected_reward: u64,
+        prev_hash: String,
+        seed_height: u64,
+        seed_hash: String,
+        next_seed_hash: String,
+        blocktemplate_blob: String,
+        blockhashing_blob: String,
+    }
+}
```

Comment:
## Type generator macro 1/3

Basically the same as before but instead of `Request {}` and `Response {}`, you specify one of the bases mentioned before. The macro picks up on this, and flattens that into your generated struct.

E.g. here it is `ResponseBase {}` so `{ status: Status, untrusted: bool }` will automatically get added to the generated `GetBlockTemplateResponse`.